File: rust/src/_347_top_k_frequent_elements.rs
/*
 * @lc app=leetcode id=347 lang=rust
 *
 * [347] Top K Frequent Elements
 */

// @lc code=start
impl Solution {
    pub fn top_k_frequent(mut nums: Vec<i32>, k: i32) -> Vec<i32> {
        nums.sort_unstable();
        let mut freqs: Vec<(usize, i32)> = nums
            .chunk_by(|a, b| a == b)
            .map(|chunk| (chunk.len(), chunk[0]))
            .collect();
        freqs.sort_unstable();
        freqs[freqs.len() - k as usize..]
            .iter()
            .map(|(_len, num)| *num)
            .collect()
    }
}
// @lc code=end

struct Solution;

File: typescript/src/_347_top_k_frequent_elements.ts
/*
 * @lc app=leetcode id=347 lang=typescript
 *
 * [347] Top K Frequent Elements
 */

// @lc code=start
function topKFrequent(nums: number[], k: number): number[] {
    const numToFreq = new Map<number, number>();
    for (const n of nums) {
        numToFreq.set(n, (numToFreq.get(n) ?? 0) + 1);
    }
    const freqToNum = new Map<number, number>();
    for (const [n, f] of numToFreq.entries()) {
        freqToNum.set(f, n);
    }
    return [...freqToNum]
        .sort()
        .reverse()
        .slice(0, k)
        .map((v, _i, _a) => v[1])
};
// @lc code=end

console.log(topKFrequent([1, 1, 1, 2, 2, 3], 2));
// [1, 2]: number[]
